keps/sig-node/1948-deallocate-dp-api/README.md

```diff
@@ -0,0 +1,854 @@
+a VM
+A device may support all of those modi or a subset. Depending on what is needed
+for the workload a device may be bound to VFIO for performance reasons e.g. and
+released/deallocated for usage in a different workload as an mediated device.
+
+- Bind/Unbind: Instead of pre-configuring the devices on the node with the
+correct drivers etc, interested in letting the device plugin handle the
+device configuration (driver bind/unbind etc). Basically, trying to
+dynamically switch the drivers bound to the device during Allocate and unbind
+the driver once the device is no longer in use. Currently no way of
+learning this state and Deallocate interface would be really useful here.
+
+- Shared Devices: The device plugin assumes the allocated devices are local to
+the node but that doesn't necessarily have to be the case. If the discovered
+devices are shared across nodes, then a `Deallocate(..)` is necessary to
+guarantee the resources are quickly freed for others to use
+(This explains [Akri](https://github.com/project-akri/akri) usecase)
+
+## Proposal
+
+The device plugin API includes API calls for:
+
+- `Allocate`: Which is used to instruct device plugins to allocate device(s) to
+requesting containers.
+- `PreStartContainer`: (Optional). Which allows the device plugins to do device
+initialization, loading drivers, and any other initialization actions that may
+be needed.
+
+This KEP proposes adding two extra API calls, maintaining the same logical
+reasoning of the previous two. Those are:
+
+- `Deallocate`: (Optional). Which is the opposite of allocate, and is needed to
+inform device plugins that some devices are no longer being used (this used to
+happen silently before)
+- `PostStopContainer`: (Optional). Which allows the device plugins to do device
+cleanup, driver unloading, and any other cleanup actions that may be needed.
+
+Since both additions are optional, existing device plugins should continue
+functioning properly with no needed modifications. Only device plugins that wish
+to utilize the new API calls will need to be modified.
+
+### Risks and Mitigations
+
+Only risk is breaking existing device plugins by introducing non optional
+changes. Can be mitigated by enough test coverage.
+
+## Design Details
+
+- Move `PreStartContainer` in `DeviceManager` to be used as a container
+lifecycle hook. (This change isn't truly required, but useful for compatibility
+and organization with the next steps).
+- Add `PostStopContainer` and `Deallocate` calls in the DevicePlugin API.
+- Add `PostStopContainer` as a container lifecycle hook.
+- Add `Deallocate` calls in container manager, taking care to only do so for
+devices that are no longer in the reuse list.
+- Add and modify test cases for both calls.
+- Test with existing device plugins to ensure the changes are non-breaking.
+- Test with new device plugins utilizing such changes to ensure the changes are
+working.
+
```

Comment:
We are missing quite a few details about how this will work here.

For example:
* What will the protobuf signatures of the new calls be?
* What happpens if the plugin is unreachable when Deallocate/PostStopContainer get called? Is it best effort, or do we need some way of making sure the plugin "finds out" about devices that have been deallocated while it was offline.
* If users are planning to start doing actual device initialization/deconfiguration as part of Allocate/Deallocate or PreStart/PostStop, how do we make sure shared devices are not reinitialized or deconfigured at an inappropriate time?

Reply:
For the second bullet, we *could* have added this info as a response to `Registration()`, but unfortunately it has a return type of `Empty` (which gets reused across a bunch of calls) so we can't just tack something onto this.

For the third bullet, I think a `refcount` attached to each device being allocated/deallocated would suffice, but we need somewhere to hang that since the protobuf messages only currently work on a repeated string type.

Reply:
> What will the protobuf signatures of the new calls be?
```
Deallocate(podUID string, containerName string)

PreStartContainer(pod *v1.Pod, container *v1.Container) error

PostStopContainer(podUID string, containerName string)
```
> What happpens if the plugin is unreachable when Deallocate/PostStopContainer get called? Is it best effort, or do we need some way of making sure the plugin "finds out" about devices that have been deallocated while it was offline.

Could we mark the device as unhealthy? If it is Allocated and Unhealthy run the Deallocate/PostStopContainer

> If users are planning to start doing actual device initialization/deconfiguration as part of Allocate/Deallocate or PreStart/PostStop, how do we make sure shared devices are not reinitialized or deconfigured at an inappropriate time?

Since the refcount is affected by Allocate/Deallocate we could add come info to the Allocate and Deallocate call.

Reply:
I don't follow what you are proposing with this:
```
Allocate(pod *v1.Pod, container *v1.Container, <refcount>)
Deallocate(podUID string, containerName string, <refcount>)

PreStartContainer(pod *v1.Pod, container *v1.Container) error
PostStopContainer(podUID string, containerName string)
```

The current API for `Allocate()`, for example is:
```
rpc Allocate(AllocateRequest) returns (AllocateResponse)

message AllocateRequest {
	repeated ContainerAllocateRequest container_requests = 1;
}

message ContainerAllocateRequest {
	repeated string devices_ids = 1 [(gogoproto.customname) = "DevicesIDs"];
}
```
Are you proposing we start embedding `pod` and `container` information in the `AllocateRequest` message (which until now we have explicitly avoided on purpose), as well as a single `refcount` of some sort. At a minimum, I think we would need a refcount per-device in the `device_ids` list, not just one per call.

I have similar concerns with the other calls you listed, i.e. *none* of them currently pass any state about which pod or container a device is destined for (on purpose), and I don't see why adding this info would be required for the purpose of this KEP. This info is purposely hidden from the plugins to keep them stateless.

Reply:
> Could we mark the device as unhealthy? If it is Allocated and Unhealthy run the Deallocate/PostStopContainer

The plugin itself is the one that track the "healthiness" of a device and informs the kubelet of this (not the other way around). The kubelet needs some API to tell the plugin which devices used to be allocated (but aren't anymore) so that it can somehow consolidate that with what devices it thinks are available/healthy/unhealthy, etc.

Reply:
> Since the refcount is affected by Allocate/Deallocate we could add some info to the Allocate and Deallocate call.

I think this info needs to somehow be attached to a device itself and passed around with it in addition to the string that the plugin announces during registration. This will come in as part of the Allocate/Deallocate calls, but it also needs to be a part of the list of devices that the kubelet "reconciles" with the plugin if it ever disappears and then comes back at a later point in time.